File: v6/tsv2/runtime/structPlane.ts
/**
 * Resolves relation-shaped values arriving in reference columns.
 *
 * The wire carries a complete target row. Before the parent tick runs, this
 * module validates the row shape, recursively resolves referenced children,
 * checks the target relation for a same-key/full-row conflict, inserts missing
 * targets, looks up their `__id`, and replaces each parent wire value with that
 * integer endpoint.
 *
 * Each target relation costs three set-based SQL statements for a non-empty
 * batch: conflict preflight, INSERT OR IGNORE, and key lookup. Statement count
 * is flat in requested row count. The target table stores typed columns plus
 * `__id`; canonical JSON exists only as transient wire and comparison text.
 *
 * Generated ticks pass an ordinary arrival applicator. Target rows therefore
 * enter the same relation clock as authored arrivals before the rewritten
 * parent rows. Direct insertion remains available to boot and focused storage
 * receipts that do not run a tick.
 */

import { concatMap, map, of, type Observable } from "rxjs";

import { TextPlane } from "./textPlane.ts";
import type {
  IArrivalBatch,
  IArrivalRow,
  IRow,
  IRowValue,
  ISqlSeam,
  IStructPlane,
  IStructTypePlan,
  IStructRefColumns,
  ITextInternPlan,
} from "./types.ts";

/** Sorted object keys, no whitespace: the ruled cross-target encoding
 *  (json_ticklog_encoding), and a clause-for-clause match of
 *  0_type_plane.pl:canonical_json_text/2. */
function canonicalize(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(canonicalize);
  if (value !== null && typeof value === "object") {
    const record = value as Record<string, unknown>;
    return Object.fromEntries(Object.keys(record).sort().map((key) => [key, canonicalize(record[key])]));
  }
  return value;
}

function canonical_text(value: unknown): string {
  return JSON.stringify(canonicalize(value));
}

function isObject(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

/** Host and HTTP arrivals carry JSON through IRowValue's text side. Schedule
 * fixtures may still inject the already-decoded object. Both spellings enter
 * the same shape check and canonicalization path before interning. */
function decoded_struct_value(value: unknown): unknown {
  if (typeof value !== "string") return value;
  try {
    return JSON.parse(value) as unknown;
  } catch {
    return value;
  }
}

/**
 * The runtime half of SLOT-ARRIVAL-MALFORMED. The oracle door checks the same
 * shapes over a fixture's static Initial + Schedule
 * (0_type_plane.pl:world_row_shape_violation/3); here the data is not static,
 * so the check runs per value at intern time and names the same reasons.
 *
 * The oracle canonicalizes object key order at ingress. This door accepts any
 * key order and canonicalizes before target lookup, producing the same row.
 */
function check_shape(plan: IStructTypePlan, by_name: ReadonlyMap<string, IStructTypePlan>, value: unknown): void {
  if (!isObject(value)) {
    throw new Error(`type_arrival_shape_mismatch: not_an_object(${plan.name}, ${JSON.stringify(value)})`);
  }
  for (const column of plan.columns) {
    if (!(column in value)) {
      throw new Error(`type_arrival_shape_mismatch: missing_key(${plan.name}, ${column})`);
    }
  }
  for (const key of Object.keys(value)) {
    if (!plan.columns.includes(key)) {
      throw new Error(`type_arrival_shape_mismatch: unknown_key(${plan.name}, ${key})`);
    }
  }
  plan.columns.forEach((column, index) => {
    const ref_type = plan.refs[index];
    const field = value[column];
    if (ref_type !== null && ref_type !== undefined) {
      const child_plan = by_name.get(ref_type);
      if (child_plan === undefined) {
        throw new Error(`type_arrival_shape_mismatch: column_type_unknown(${ref_type})`);
      }
      check_shape(child_plan, by_name, field);
    }
  });
}

/** `TypeName` ++ canonical JSON. The type name is an identifier and the
 *  rendering always starts with `{`, so the concatenation is unambiguous, and
 *  it matches lower.pl:struct_intern_statements/5 byte for byte. Full
 *  canonical text rather than a digest because @libsql registers no UDF and
 *  SQLite ships no hash (SLOT-SEMANTIC-DIGEST in 0_type_plane.pl's header):
 *  strictly stronger than a hash, with no collision case to reason about. */
function semantic_key(type_name: string, rendered: string): string {
  return `${type_name}${rendered}`;
}

interface ICollected {
  readonly rendered: string;
  /** column values; a ref column holds the CHILD'S semantic key until the
   *  child's own type has been interned and its dense id is known. */
  readonly fields: readonly (IRowValue | { readonly child_semantic: string })[];
}

function collect(
  plan: IStructTypePlan,
  by_name: ReadonlyMap<string, IStructTypePlan>,
  value: unknown,
  per_type: Map<string, Map<string, ICollected>>,
): string {
  const decoded = decoded_struct_value(value);
  check_shape(plan, by_name, decoded);
  const object = decoded as Record<string, unknown>;
  const fields = plan.columns.map((column, index) => {
    const ref_type = plan.refs[index];
    if (ref_type === null || ref_type === undefined) return object[column] as IRowValue;
    // Post-order: the child is collected (and therefore interned) first.
    const child_semantic = collect(by_name.get(ref_type)!, by_name, object[column], per_type);
    return { child_semantic };
  });
  const rendered = canonical_text(object);
  const semantic = semantic_key(plan.name, rendered);
  let bucket = per_type.get(plan.name);
  if (bucket === undefined) {
    bucket = new Map<string, ICollected>();
    per_type.set(plan.name, bucket);
  }
  bucket.set(semantic, { rendered, fields });
  return semantic;
}

function rewrite_row(
  row: IRow,
  refs: readonly (string | null)[],
  by_name: ReadonlyMap<string, IStructTypePlan>,
  ids: ReadonlyMap<string, number>,
): IRow {
  return row.map((value, index) => {
    const ref_type = refs[index];
    if (ref_type === null || ref_type === undefined) return value;
    const rendered = canonical_text(decoded_struct_value(value));
    const id = ids.get(semantic_key(ref_type, rendered));
    if (id === undefined) {
      throw new Error(`relation reference normalization lost the id for ${ref_type} value ${rendered}`);
    }
    return id;
  });
}

export const StructPlane: IStructPlane = {
  canonical_text,

  intern(
    seam: ISqlSeam,
    types: readonly IStructTypePlan[],
    ref_columns: IStructRefColumns,
    arrivals: IArrivalBatch,
    apply_targets?: (arrivals: IArrivalBatch) => Observable<unknown>,
    text_plan?: ITextInternPlan,
  ): Observable<IArrivalBatch> {
    if (types.length === 0 || arrivals.length === 0) return of(arrivals);
    const by_name = new Map(types.map((plan) => [plan.name, plan]));
    const per_type = new Map<string, Map<string, ICollected>>();
    for (const arrival of arrivals) {
      const refs = ref_columns[arrival.rel];
      if (refs === undefined) continue;
      arrival.row.forEach((value, index) => {
        const ref_type = refs[index];
        if (ref_type === null || ref_type === undefined) return;
        collect(by_name.get(ref_type)!, by_name, value, per_type);
      });
    }
    if (per_type.size === 0) return of(arrivals);

    // `types` arrives in topological order (lower.pl:struct_type_plans/2), so
    // one left fold down the list resolves every child before its parent.
    const ids = new Map<string, number>();
    const pending = types.filter((plan) => per_type.has(plan.name));
    return pending.reduce<Observable<unknown>>(
      (chain, plan) => chain.pipe(concatMap(() =>
        intern_one_type(seam, plan, per_type.get(plan.name)!, ids, apply_targets, text_plan)
      )),
      of(undefined),
    ).pipe(
      map(() => arrivals.map((arrival): IArrivalRow => {
        const refs = ref_columns[arrival.rel];
        if (refs === undefined) return arrival;
        return { rel: arrival.rel, sign: arrival.sign, row: rewrite_row(arrival.row, refs, by_name, ids) };
      })),
    );
  },
};

function intern_one_type(
  seam: ISqlSeam,
  plan: IStructTypePlan,
  bucket: ReadonlyMap<string, ICollected>,
  ids: Map<string, number>,
  apply_targets: ((arrivals: IArrivalBatch) => Observable<unknown>) | undefined,
  text_plan: ITextInternPlan | undefined,
): Observable<unknown> {
  const semantics = [...bucket.keys()];
  const rows = [...bucket.values()].map((collected) =>
    collected.fields.map((field) => (is_child_reference(field) ? id_for(ids, field.child_semantic) : field)) as IRow
  );
  const arrivals: IArrivalBatch = rows.map((row): IArrivalRow => ({
    rel: plan.name,
    sign: "add",
    row,
  }));
  // The arrival door never sees a target row, and the preflight, insert and
  // key lookup all read one tuple, so its text columns intern before encoding.
  const staged = text_plan === undefined
    ? of(arrivals)
    : TextPlane.intern(seam, text_plan, arrivals);
  return staged.pipe(
    concatMap((interned) => intern_target_rows(seam, plan, semantics, interned, ids, apply_targets)),
  );
}

function intern_target_rows(
  seam: ISqlSeam,
  plan: IStructTypePlan,
  semantics: readonly string[],
  arrivals: IArrivalBatch,
  ids: Map<string, number>,
  apply_targets: ((arrivals: IArrivalBatch) => Observable<unknown>) | undefined,
): Observable<unknown> {
  const lookup_to_semantic = new Map<string, string>();
  const tuple_by_key = new Map<string, string>();
  const tuples = arrivals.map((arrival, index) => {
    const fields = arrival.row;
    const tuple = JSON.stringify(fields);
    const key = JSON.stringify(plan.key_indices.map((position) => fields[position]));
    const prior = tuple_by_key.get(key);
    if (prior !== undefined && prior !== tuple) {
      throw new Error(`relation_reference_conflict(${plan.name}, ${key}, ${prior}, ${tuple})`);
    }
    tuple_by_key.set(key, tuple);
    lookup_to_semantic.set(tuple, semantics[index]!);
    return fields;
  });
  const encoded = JSON.stringify(tuples);
  return seam.runner.execute(seam.db, { sql: plan.conflict_sql, args: [encoded] }).pipe(
    map((result) => {
      if (result.rows.length === 0) return undefined;
      const row = result.rows[0]!;
      throw new Error(
        `relation_reference_conflict(${plan.name}, ${String(row["__requested"])}, ${String(row["__stored"])})`,
      );
    }),
    concatMap(() => {
      return apply_targets === undefined
        ? seam.runner.execute(seam.db, { sql: plan.intern_sql, args: [encoded] })
        : apply_targets(arrivals);
    }),
    concatMap(() => seam.runner.execute(seam.db, { sql: plan.lookup_sql, args: [encoded] })),
    map((result) => {
      for (const row of result.rows) {
        const lookup = row["__lookup"] as string;
        const stored = row["__stored"] as string;
        if (stored !== lookup) {
          throw new Error(`relation_reference_conflict(${plan.name}, ${lookup}, ${stored})`);
        }
        const semantic = lookup_to_semantic.get(lookup);
        if (semantic === undefined) {
          throw new Error(`relation reference lookup returned an unknown row ${String(row["__lookup"])}`);
        }
        ids.set(semantic, Number(row["__id"]));
      }
      return undefined;
    }),
  );
}

/** IRowValue already admits a plain object, so `"child_semantic" in field`
 *  alone leaves the field's own type in play; the predicate is what narrows. */
function is_child_reference(
  field: IRowValue | { readonly child_semantic: string },
): field is { readonly child_semantic: string } {
  return (
    typeof field === "object" && field !== null && !Array.isArray(field) &&
    "child_semantic" in field && typeof field.child_semantic === "string"
  );
}

function id_for(ids: ReadonlyMap<string, number>, semantic: string): number {
  const id = ids.get(semantic);
  if (id === undefined) {
    throw new Error(`relation reference normalization read a child id before its target row: ${semantic}`);
  }
  return id;
}

File: v6/sprefa-engine-rs/src/struct_plane.rs
use std::collections::HashMap;

use serde_json::Value as JsonValue;

use crate::sql::{column_index, SqlRunner, SqliteSeam};
use crate::types::{
    Arrival, ArrivalSign, BoundaryResult, IncrementalRelationPlan, Row, ScalarValue, SqlStatement,
    StructTypePlan, TextInternPlan, Value,
};

#[derive(Clone)]
enum CollectedField {
    Scalar(Value),
    Child(String),
}

#[derive(Clone)]
struct Collected {
    semantic: String,
    fields: Vec<CollectedField>,
}

fn parse_struct(value: &Value) -> JsonValue {
    let Value::Text(text) = value else {
        panic!("type_arrival_shape_mismatch: not_an_object")
    };
    serde_json::from_str(text)
        .unwrap_or_else(|_| panic!("type_arrival_shape_mismatch: not_an_object"))
}

fn canonical_text(value: &JsonValue) -> String {
    serde_json::to_string(value).expect("struct value serialization failed")
}

fn semantic_key(type_name: &str, rendered: &str) -> String {
    format!("{}{}", type_name, rendered)
}

fn json_scalar(value: &JsonValue) -> Value {
    match value {
        JsonValue::Number(number) => match number.as_i64() {
            Some(integer) => Value::Integer(integer),
            None => Value::Real(number.as_f64().expect("struct number is outside f64")),
        },
        JsonValue::Bool(flag) => Value::Bool(*flag),
        JsonValue::String(text) => Value::Text(text.clone()),
        JsonValue::Null => Value::Text(String::new()),
        JsonValue::Array(_) | JsonValue::Object(_) => Value::Text(canonical_text(value)),
    }
}

type CollectedBucket = (Vec<Collected>, HashMap<String, usize>);

fn collect(
    plan: &StructTypePlan,
    by_name: &HashMap<&str, &StructTypePlan>,
    value: &JsonValue,
    per_type: &mut HashMap<String, CollectedBucket>,
) -> String {
    let object = value
        .as_object()
        .unwrap_or_else(|| panic!("type_arrival_shape_mismatch: not_an_object({})", plan.name));
    for column in &plan.columns {
        if !object.contains_key(column) {
            panic!(
                "type_arrival_shape_mismatch: missing_key({}, {})",
                plan.name, column
            );
        }
    }
    for key in object.keys() {
        if !plan.columns.contains(key) {
            panic!(
                "type_arrival_shape_mismatch: unknown_key({}, {})",
                plan.name, key
            );
        }
    }
    let fields = plan
        .columns
        .iter()
        .enumerate()
        .map(|(index, column)| {
            let field = &object[column];
            match plan.refs.get(index).and_then(Option::as_deref) {
                None => CollectedField::Scalar(json_scalar(field)),
                Some(child_name) => {
                    let child = by_name.get(child_name).unwrap_or_else(|| {
                        panic!("type_arrival_shape_mismatch: column_type_unknown({child_name})")
                    });
                    CollectedField::Child(collect(child, by_name, field, per_type))
                }
            }
        })
        .collect();
    let rendered = canonical_text(value);
    let semantic = semantic_key(&plan.name, &rendered);
    let bucket = per_type.entry(plan.name.clone()).or_default();
    let collected = Collected {
        semantic: semantic.clone(),
        fields,
    };
    COLLECT_PROBES.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
    match bucket.1.get(&semantic) {
        Some(index) => bucket.0[*index] = collected,
        None => {
            bucket.1.insert(semantic.clone(), bucket.0.len());
            bucket.0.push(collected);
        }
    }
    semantic
}

/// One comparison per collected value under an index, one per pair under a scan.
static COLLECT_PROBES: std::sync::atomic::AtomicU64 = std::sync::atomic::AtomicU64::new(0);

pub fn collect_probes() -> u64 {
    COLLECT_PROBES.load(std::sync::atomic::Ordering::Relaxed)
}

fn resolve_fields(collected: &Collected, ids: &HashMap<String, i64>) -> Row {
    collected
        .fields
        .iter()
        .map(|field| match field {
            CollectedField::Scalar(value) => value.clone(),
            CollectedField::Child(semantic) => Value::Integer(
                *ids.get(semantic)
                    .unwrap_or_else(|| panic!("relation reference child id missing: {semantic}")),
            ),
        })
        .collect()
}

fn encoded_rows(rows: &[Row]) -> BoundaryResult<ScalarValue> {
    let encoded = rows
        .iter()
        .map(|row| crate::incremental::json_array_text(row))
        .collect::<BoundaryResult<Vec<_>>>()?
        .join(",");
    Ok(ScalarValue::Text(format!("[{}]", encoded)))
}

fn intern_type(
    work: &crate::incremental::TickWork,
    seam: &SqliteSeam,
    plan: &StructTypePlan,
    collected: &[Collected],
    ids: &mut HashMap<String, i64>,
    relations: &[IncrementalRelationPlan],
    text_plan: Option<&TextInternPlan>,
) -> BoundaryResult<()> {
    let rows: Vec<Row> = collected
        .iter()
        .map(|entry| resolve_fields(entry, ids))
        .collect();
    let arrivals: Vec<Arrival> = rows
        .iter()
        .map(|row| Arrival {
            rel: plan.name.clone(),
            sign: ArrivalSign::Add,
            row: row.clone(),
        })
        .collect();
    let staged = match text_plan {
        Some(text_plan) => {
            crate::text_plane::intern(seam, text_plan, std::borrow::Cow::Owned(arrivals))?
        }
        None => std::borrow::Cow::Owned(arrivals),
    };
    let staged_rows: Vec<Row> = staged.iter().map(|arrival| arrival.row.clone()).collect();
    let encoded = encoded_rows(&staged_rows)?;
    let conflict = seam
        .execute(&SqlStatement {
            sql: plan.conflict_sql.clone(),
            args: vec![encoded.clone()],
        })
        .expect("relation reference conflict query failed");
    if !conflict.rows.is_empty() {
        panic!("relation_reference_conflict({})", plan.name);
    }
    crate::incremental::apply_arrivals(seam, &staged, relations, work)?;
    let lookup = seam
        .execute(&SqlStatement {
            sql: plan.lookup_sql.clone(),
            args: vec![encoded],
        })
        .expect("relation reference lookup failed");
    let lookup_index = column_index(&lookup, "__lookup").expect("struct lookup column missing");
    let id_index = column_index(&lookup, "__id").expect("struct id column missing");
    let mut semantics_by_tuple: HashMap<String, &str> = HashMap::new();
    for (entry, row) in collected.iter().zip(staged_rows.iter()) {
        semantics_by_tuple.insert(crate::incremental::json_array_text(row)?, &entry.semantic);
    }
    for row in &lookup.rows {
        let tuple = match row.get(lookup_index) {
            Some(Value::Text(tuple)) => tuple,
            _ => panic!("struct lookup tuple missing"),
        };
        let id = row
            .get(id_index)
            .and_then(Value::as_i64)
            .expect("struct lookup id missing");
        let semantic = semantics_by_tuple
            .get(tuple)
            .unwrap_or_else(|| panic!("relation reference lookup returned an unknown row {tuple}"));
        ids.insert((*semantic).to_string(), id);
    }
    Ok(())
}

fn rewrite_row(row: &Row, refs: &[Option<String>], ids: &HashMap<String, i64>) -> Row {
    row.iter()
        .enumerate()
        .map(
            |(index, value)| match refs.get(index).and_then(Option::as_deref) {
                None => value.clone(),
                Some(type_name) => {
                    let decoded = parse_struct(value);
                    let rendered = canonical_text(&decoded);
                    let semantic = semantic_key(type_name, &rendered);
                    Value::Integer(*ids.get(&semantic).unwrap_or_else(|| {
                        panic!("relation reference normalization lost the id for {semantic}")
                    }))
                }
            },
        )
        .collect()
}

/// The distinct struct values one batch names, per type name. Public so a
/// COUNT test can read `collect_probes` over it without a seam.
pub fn distinct_struct_values(
    types: &[StructTypePlan],
    ref_columns: &HashMap<String, Vec<Option<String>>>,
    arrivals: &[Arrival],
) -> usize {
    let by_name: HashMap<&str, &StructTypePlan> = types
        .iter()
        .map(|plan| (plan.name.as_str(), plan))
        .collect();
    collect_arrivals(&by_name, ref_columns, arrivals)
        .values()
        .map(|bucket| bucket.0.len())
        .sum()
}

fn collect_arrivals(
    by_name: &HashMap<&str, &StructTypePlan>,
    ref_columns: &HashMap<String, Vec<Option<String>>>,
    arrivals: &[Arrival],
) -> HashMap<String, CollectedBucket> {
    let mut per_type: HashMap<String, CollectedBucket> = HashMap::new();
    for arrival in arrivals.iter() {
        let Some(refs) = ref_columns.get(&arrival.rel) else {
            continue;
        };
        for (index, value) in arrival.row.iter().enumerate() {
            let Some(type_name) = refs.get(index).and_then(Option::as_deref) else {
                continue;
            };
            let plan = by_name
                .get(type_name)
                .unwrap_or_else(|| panic!("struct type plan missing: {type_name}"));
            let decoded = parse_struct(value);
            collect(plan, by_name, &decoded, &mut per_type);
        }
    }
    per_type
}

pub fn intern<'a>(
    seam: &SqliteSeam,
    types: &[StructTypePlan],
    ref_columns: &HashMap<String, Vec<Option<String>>>,
    arrivals: std::borrow::Cow<'a, [Arrival]>,
    relations: &[IncrementalRelationPlan],
    text_plan: Option<&TextInternPlan>,
    work: &crate::incremental::TickWork,
) -> BoundaryResult<std::borrow::Cow<'a, [Arrival]>> {
    if types.is_empty() || arrivals.is_empty() {
        return Ok(arrivals);
    }
    let by_name: HashMap<&str, &StructTypePlan> = types
        .iter()
        .map(|plan| (plan.name.as_str(), plan))
        .collect();
    let per_type = collect_arrivals(&by_name, ref_columns, &arrivals);
    if per_type.is_empty() {
        return Ok(arrivals);
    }
    let mut ids = HashMap::new();
    for plan in types {
        if let Some(collected) = per_type.get(&plan.name) {
            intern_type(
                work,
                seam,
                plan,
                &collected.0,
                &mut ids,
                relations,
                text_plan,
            )?;
        }
    }
    let rewritten: Vec<Arrival> = arrivals
        .iter()
        .map(|arrival| match ref_columns.get(&arrival.rel) {
            None => arrival.clone(),
            Some(refs) => Arrival {
                rel: arrival.rel.clone(),
                sign: arrival.sign,
                row: rewrite_row(&arrival.row, refs, &ids),
            },
        })
        .collect();
    Ok(std::borrow::Cow::Owned(rewritten))
}
